crates/edr_eth/src/receipt.rs

```diff
@@ -10,57 +10,84 @@ mod block;
 mod transaction;
 
 use alloy_rlp::{Buf, BufMut, Decodable, Encodable};
+use revm_primitives::SpecId;
+use serde::{ser::SerializeStruct, Serialize, Serializer};
 
 pub use self::{block::BlockReceipt, transaction::TransactionReceipt};
-use crate::{Bloom, B256};
+use crate::{Bloom, B256, U64};
 
 /// Typed receipt that's generated after execution of a transaction.
-#[derive(Clone, Debug, PartialEq, Eq)]
-#[cfg_attr(feature = "serde", derive(serde::Serialize))]
-#[cfg_attr(feature = "serde", serde(deny_unknown_fields))]
-#[cfg_attr(feature = "serde", serde(rename_all = "camelCase"))]
+#[derive(Clone, Debug)]
 pub struct TypedReceipt<LogT> {
     /// Cumulative gas used in block after this transaction was executed
-    #[cfg_attr(feature = "serde", serde(with = "crate::serde::u64"))]
     pub cumulative_gas_used: u64,
     /// Bloom filter of the logs generated within this transaction
     pub logs_bloom: Bloom,
     /// Logs generated within this transaction
     pub logs: Vec<LogT>,
-    #[cfg_attr(feature = "serde", serde(flatten))]
+    /// The typed receipt data.
+    /// - `root` field (before Byzantium) or `status` field (after Byzantium)
+    /// - `type` field after Berlin
     pub data: TypedReceiptData,
+    /// The currently active hardfork in the local blockchain. Hack for
+    /// serialization. Not included in the serialized representation.
+    /// Assumes remote runs latest hardfork.
+    pub spec_id: SpecId,
```

Comment:
We need the hardfork for serialization as the `type` field should be only present after Berlin. This requires the assumption that the remote runs the latest hardfork and that we only deserialize from a remote. Alternative implementation without the assumption would be to have two types: one for serialization containing the spec id and one for deserializatin without the spec id. I started doing the alternative implementation first, but it required two many changes, that's why I ended up with the current one.

Reply:
Was it not possible to make the field in question optional for (de)serialization purposes?